### ISSUE TITLE
Policy key already exists

### ISSUE THREAD

**Issue by BernieWhite** [Azure/PSRule.Rules.Azure#1653]
**Description of the issue**

When policy definitions are assigned multiple times (which may have different parameters) the subsequent occurrences fail.

**Expected behaviour**

We need to:

- [x] Ignore adding cases were the same policy definition is assigned with the same parameters to prevent duplicates.
- [x] Generate a unique `name` and `ref` for each policy rule. We should be able to do this via a hash of the condition/ pre-conditions.

For example:

- Ref: `Policy.e749c2d003da`

**Error output**

```text
Export-AzPolicyAssignmentRuleData: An item with the same key has already been added. Key: /providers/Microsoft.Authorization/policyDefinitions/26a828e1-e88f-464e-bbb3-c134a282b9de
```

**Module in use and version:**

- Module: PSRule.Rules.Azure
- Version: **1.20.0-B0028**

Captured output from `$PSVersionTable`:

```text
Name                           Value
----                           -----
PSVersion                      7.2.6
PSEdition                      Core
GitCommitId                    7.2.6
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Additional context**

Related to #181

**Comment by BernieWhite** [comment]
@fbinotto I think this issue will be dealing with the dictionaries defined [here](https://github.com/Azure/PSRule.Rules.Azure/blob/71b0beadd431981aae24226016b52950bac7006f/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs#L86-L92) when new items are added.

**Comment by ArmaanMcleod** [comment]
@BernieWhite I can work on this one.

Not sure how we would generate a unique `name` for the policy rules, since the current implementation just uses the definition name. Could we keep the same `name` but ensure different `ref`? 

Also, not sure how we'd handle cases where the same policy definition is assigned but different variations of parameters are used. Would we include all rules and generate unique `ref` for each? For assignments with the same parameters, we can skip those.

**Comment by BernieWhite** [comment]
@ArmaanMcleod `name` and `ref` both need to be unique.

Yes. I think the error is a dictionary collision, however I think we are still going to need to resolve it more broadly.

---

Consider an initiative assignment that might be used for tagging. To enforce two or more tags it's highly likely and recommended that a customer would reference the same policy definition twice such as the built-in `Require a tag on resource groups` or `Inherit a tag from the resource group`.

In both cases the customer would just use the `tagName` parameter to supply the actual tag.

---

I was thinking a hash could be generated of the `Condition` property here: https://github.com/Azure/PSRule.Rules.Azure/blob/7be571cee4fe9b3423add5e25b2417df43bc3a12/src/PSRule.Rules.Azure/Data/Policy/Models.cs#L105

Although we would probably also need to think about any pre-conditions that would come out of #1708 at some point.

If we generate a hash based on _definitionId_ + _condition_ + _pre-conditions_ this should give us something unique that can be used to avoid collisions and deterministic.

**Comment by ArmaanMcleod** [comment]
@BernieWhite I'm assuming your tags example would like:

```json
"policyDefinitions": [
      {
        "policyDefinitionReferenceId": "Inherit a tag from the subscription_1",
        "policyDefinitionId": "/providers/Microsoft.Authorization/policyDefinitions/b27a0cbd-a167-4dfa-ae64-4337be671140",
        "parameters": {
          "tagName": {
            "value": "costCentre"
          }
        },
        "groupNames": []
      },
      {
        "policyDefinitionReferenceId": "Inherit a tag from the subscription_2",
        "policyDefinitionId": "/providers/Microsoft.Authorization/policyDefinitions/b27a0cbd-a167-4dfa-ae64-4337be671140",
        "parameters": {
          "tagName": {
            "value": "department"
          }
        },
        "groupNames": []
      }
    ]
```

The problem I see here is `Export-AzPolicyAssignmentData` won't include these `tagName` parameter values.

**Comment by BernieWhite** [comment]
@ArmaanMcleod That's one variation, but definition parameter can be fed from initiative assignment parameters as well.

`policyDefinitionReferenceId` is only unique within an initiative nor specifically between initiatives, not sure on what the value including it yet would be.

Using the `Condition` has advantages because it is already processed by the export process it is going to help deduplicate rules. You could do this via the unprocessed policy definitions but it would be likely more work and more complex.

**Comment by ArmaanMcleod** [comment]
@BernieWhite Yeah fully agree, I think the issue I'm seeing is how the visitor is processing parameters before expanding the policy rule: https://github.com/Azure/PSRule.Rules.Azure/blob/main/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs#L171-L182

In order for it to look up and expand parameters, the parameters from assignment level are set beforehand. This means for the condition approach to be reliable, it must be processed and expanded first, otherwise policy definitions will be duplicated since the conditions will remain the same.

Gets challenging to do definition duplication with condition hash before the condition is expanded 😄.

**Comment by BernieWhite** [comment]
@ArmaanMcleod I've refactored for a fair bit for #1708. I think it should be easier to implement this with the latest updates.

If you get stuck, let me know.

**Comment by ArmaanMcleod** [comment]
@BernieWhite What do you think of this for unique hashes for `ref` and `name`:

```json
"name": "Azure.Policy.f5388351250d",
"ref": "Azure.Policy.f5388351250d",
```

I kept them the same for now as they both use the hash of the definitionId + condition + pre-condition.

**Comment by BernieWhite** [comment]
@ArmaanMcleod That should be fine. In terms of `Azure.` I was thinking that customers may want to customize this prefix and it may make sense if customers have multiple scopes or environments that want to run tests for, but it could default to `Azure.`.

Configuration would be by `ps-rule.yaml` or command-line parameter.

Maybe let's allow for this. And either tackle this in this or a separate issue.

**Comment by ArmaanMcleod** [comment]
@BernieWhite That sounds good. Should we create something like a `PolicyRuleOption` to encapsulate these specific configurations for policy rules?

Something like this:

```yaml
configuration:
  AZURE_POLICY_RULE:
    prefix: 'something'
```

Or would you rather just keep it simpler

```yaml
configuration:
  AZURE_POLICY_RULE_PREFIX: 'something'
```

I was thinking if we should use the first approach in case we need to configure more stuff together like the other options.

**Comment by BernieWhite** [comment]
@ArmaanMcleod Let's go with option 2.
